### ISSUE TITLE
DLApplicationSpecific removed issue (ASN.1)

### ISSUE THREAD

**Issue by WalterYongtaoWang** [bcgit/bc-java#1551]
Hi,

My application has been working fine with bcutil-jdk15to18-166.jar. Since ANS1ApplicationSpecific and DLApplicationSpecific were removed in 177, I'm trying to find another way to decode the Ticket type in RFC 4120 with the newest bcutil-jdk15to18-177.jar :

* Ticket               ::= [APPLICATION 1] SEQUENCE {
 *      tkt-vno         [0] INTEGER (5),
 *      realm           [1] Realm,
 *      sname           [2] PrincipalName,
 *      enc-part        [3] EncryptedData -- EncTicketPart

ASN1Primitive prim = ASN1Primitive.fromByteArray(expected);
DERTaggedObject dto = (DERTaggedObject)prim;

Then it caused a casting issue:

java.lang.ClassCastException: class org.bouncycastle.asn1.DLApplicationSpecific cannot be cast to class org.bouncycastle.asn1.DERTaggedObject (org.bouncycastle.asn1.DLApplicationSpecific and org.bouncycastle.asn1.DERTaggedObject are in unnamed module of loader 'app')
at net.protocol.kerberos.TestTicket.test(TestTicket.java:92)

Please help:

1. How should we do this without DLApplicationSpecific?

2. The prim is an instance of DLApplicationSpecific from the debugging, and the error message also says it's DLApplicationSpecific, but this is already removed and I couldn't find this class in the bcutil-jdk15to18-177.jar either.

Really appreciate your help.

Regards,

Walter Wang

**Comment by peterdettman** [comment]
1. All tagged objects are now handled through ASN1TaggedObject, which now knows about tag class in addition to tag number. The following code should let you get to the SEQUENCE while checking the tagging is valid:

```
ASN1TaggedObject ato = ASN1TaggedObject.getInstance(prim, BERTags.APPLICATION, 1);
ASN1Sequence seq = (ASN1Sequence)ato.getBaseUniversal(true, BERTags.SEQUENCE);
```

2. The only explanation would be that you somehow still have an older jar present in the classpath somehow; DLApplicationSpecific has been completely removed from the new jars.

**Comment by WalterYongtaoWang** [comment]
Thanks a lot for the help. There was a jar conflict in my projects,  that works perfectly.

Now I have a compatibility issue with EXPLICIT/ IMPLICIT when updating from 166.

For example, I was getting the  tkt-vno with the following code with 166 and older versions:

		ASN1Integer tktVno = ASN1Integer.getInstance(tktVnoTaggedObject, false);

I have to use false for the second EXPLICIT argument and I thought that is wrong, but I had to use false to make it work.

Now, I have to use "true" to make it work, which means basically I need to change my code every where.

Was that a bug and fixed? Is there a better way to do this?

Sorry for this kind of questions, I searched but found no answers.

Appreciate your help.

**Comment by WalterYongtaoWang** [comment]
Sorry, just double tested it again with  166, bot "true" or "false" works. I believe only "false" works with some older versions.

**Comment by peterdettman** [comment]
Yes, recent versions are more strict about making sure the explicit/implicit declaration in the type matches the actual encoding. I am not aware that it was ever necessary to pass the "wrong" value (apart from perhaps bugs in specific types), but there are ASN.1 rules that can be surprising when trying to work out whether an explicit or implicit tag is expected.  e.g. values of a CHOICE type are _always_ explicitly tagged. Also be aware that the default tagging is configured by the ASN.1 module definitions (e.g. in RFC 4120 we see "DEFINITIONS EXPLICIT TAGS"), so it is not possible to just look at ASN.1 types in isolation.

**Comment by WalterYongtaoWang** [comment]
Thanks, appreciate the help.
